Update webhook-consumer.md - Round #1 (#26)
SymfonyCasts/lemon-squeezy: sfcasts/webhook-consumer.md

The *goal* is to share the user ID with LemonSqueezy when a customer places an order. If we head back to the `OrderController`, you may notice that PhpStorm isn't super happy about this `createCheckoutUrl()` call. That's because we still need to make it a requirement in that method's signature. Remove this `?` and that should do it - no need to call `denyAccessUnlessGranted()`.

To confirm everything's working as expected, over on our site, log out... add a product to the cart, and try to check out. Since we're a non-authenticated user, we *should* be redirected to the login page, and... perfect!

## Handling the Webhook Consumer

Back in the code, in our `consume()` method, set `$payload = $event->getPayload()`. Below that, say `$userId = $payload['meta']['custom_data']['user_id'] ?? null`. We can also leave a comment here mentioning that `getUser()` won't work in webhooks as a non-authenticated user in this process.

Now, let's conduct a sanity check with `if (!$userId)`. If this check fails, we'll `throw new InvalidArgumentException()` with a `sprintf()` inside saying `'User ID not found in LemonSqueezy webhook: %s', $userId`.

> To access the `EntityManger`, in our constructor, inject `private EntityManagerInterface` as `$entityManager`.

Reviewer: ```suggestion
To access the `EntityManager`, in our constructor, inject `private EntityManagerInterface $entityManager`.
```

Back in the `consume()` method, continue with `$user = $this->entityManager->getRepository(User::class)->find($userId)`. 

> Next, if `$user` doesn't exist, we'll `throw new EntityNotFoundException()` (choose the one from "Doctrine\ORM"). We'll also add `sprintf()` as an argument, stating `User "%s" not found for LemonSqueezy webhook "%s"!`, and pass `$userId` and `$event->getId()`. 

Reviewer: ```suggestion
Next, if `$user` doesn't exist, we'll `throw new EntityNotFoundException()` (choose the one from `Doctrine\ORM`). We'll also add `sprintf()` as an argument, stating `User "%s" not found for LemonSqueezy webhook "%s"!`, and pass `$userId` and `$event->getId()`. 
```

Below that, add `match ($event->getName())`, and for `order_created`, call `$this->handleOrderCreatedEvent()`. This method doesn't exist yet, but we'll create it later. Also pass `$event` and `$user` as arguments. At this point, we should only have supported events, but on the off chance we're missing something, add a `default` that will `throw new LogicException()`, with `sprintf('Unsupported LemonSqueezy event: %s', $event->getName())`. *Nice*.

## Creating the HandleOrder Event

Before we forget, let's circle back and create the `handleOrderCreatedEvent()`. This will be a `private function`, and it looks like PhpStorm added one argument - `RemoteEvent $event` - but forgot the second, so we'll add `User $user` manually.

> Inside, let's fetch the payload with `$payload = $event->getPayload()`. Below that, fetch the customer ID from the payload: `$customerId = $payload['data']['attributes']['customer_id']`. If you're wondering where this came from, you can find this path in the Ngrok request summary.

Reviewer: ```suggestion
Inside, let's fetch the payload with `$payload = $event->getPayload()`. Below that, fetch the customer ID from the payload: `$customerId = $payload['data']['attributes']['customer_id']`. If you're wondering where this came from, you can find this path in the Ngrok request payload.
```

Okay, we have the `customer_id` now, but we still need a new property on the user to save it. At your terminal, create a new tab and run:

```terminal
bin/console make:entity
```

> We'll call this `User`. For the property name, call it `lsCustomerId`. Make it a string, with a length of 255, and make it nullable. Hit "enter" one more time and... done!

Reviewer: ```suggestion
For the class name, we'll write `User`. For the property name, call it `lsCustomerId`. Make it a string, with a length of 255, and make it nullable. Hit `Enter` one more time and... done!
```

Let's log in again, add a product to the cart, and try to check out. Oops, *another* error. Let's head back to our code and uncomment this `dd()` to get a better picture of what's happening here. If we refresh our site... ah! It looks like the field needs to be a string, and it's pointing to the custom `user_id` we added. You can't pass just *anything*, so passing the whole user object definitely won't work. Head back to our code and comment out that `dd()` again... then, up here... let's specify that this user ID is a `string` and try this again. When we refresh... success! We're on the checkout page!

Let's fill in the card info... address... make the payment, and wait for the webhook. *Yes*! Our transaction was accepted and, over here, we have a 202 status code.

> If we look at the request, we can see our `custom_data` where the `user_id` equals `1`. We can also check the database with a handy dandy SQL command. At your terminal, run
> 
> ```terminal
> bin/console doctrine:query:sql
> ```
> 
> and in double quotes, input

Reviewer: ```suggestion
If we look at the request, we can see our `custom_data` where the `user_id` equals `1`. We can also check the database with a handy dandy SQL command. At your terminal, run `bin/console doctrine:query:sql` and in double quotes, input `SELECT * FROM user WHERE id = 1`:
```

> ```terminal

Reviewer: ```suggestion
```terminal-silent
```

> SELECT * FROM user WHERE id = 1

Reviewer: ```suggestion
bin/console doctrine:query:sql "SELECT * FROM user WHERE id = 1"
```

```

> since we have an ID of "1" for the current user. Hit "enter" and... *yes*! The `lsCustomerId` is set to this unique ID. This is what we need to create a list of orders made by this customer.

Reviewer: ```suggestion
since we have an ID of "1" for the current user. Hit `Enter` and... *yes*! The `lsCustomerId` is set to this unique ID. This is what we need to create a list of orders made by this customer.
```
